src/main/scala/com/fulcrumgenomics/bam/SetMateInformation.scala

```diff
@@ -0,0 +1,57 @@
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{Io, ProgressLogger}
+import dagr.commons.CommonsDef.{PathToBam, PathToFasta}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.cmdline.ValidationException
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.SamPairUtil.SetMateInfoIterator
+import htsjdk.samtools.{SAMFileWriterFactory, SamReaderFactory}
+
+import scala.collection.JavaConversions.asScalaIterator
+
+
+@clp(group=ClpGroups.SamOrBam, description=
+"""
+  |Adds and/or fixes mate information on paired-end reads. Sets the MQ (mate mapping quality),
+  |MC (mate cigar string), ensures all mate-related flag fields are set correctly, and that
+  |the mate reference and mate start position are correct.
+  |
+  |Supplementary records are handled correctly (updated with their mate's non-supplemental
+  |attributes).  Secondary alignments are passed through but are not updated.
+  |
+  |The input file must be query-name sorted or query-name grouped (i.e. all records from the same
+  |query sequence must be adjacent in the file, though the ordering between queries is unspecified).
+""")
+class SetMateInformation
+(
+  @arg(flag="i", doc="Input SAM/BAM/CRAM file.")                      val input: PathToBam = Io.StdIn,
+  @arg(flag="o", doc="Output SAM/BAM/CRAM file.")                     val output: PathToBam = Io.StdOut,
+  @arg(flag="r", doc="Reference fasta, only needed if writing CRAM.") val ref: Option[PathToFasta] = None,
+  @arg(flag="x", doc="If specified, do not fail when reads marked as paired are missing their mate pairs.")
+                                                                      val allowMissingMates: Boolean = false
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  private val in = SamReaderFactory.make().open(input.toFile)
+  if (in.getFileHeader.getSortOrder != SortOrder.queryname && in.getFileHeader.getGroupOrder != GroupOrder.query) {
+    in.close()
+    throw new ValidationException("Input is not queryname sorted or grouped.")
+  }
+
+  override def execute(): Unit = {
+    val out = new SAMFileWriterFactory().makeWriter(in.getFileHeader, true, output.toFile, ref.map(_.toFile).orNull)
+    val progress = new ProgressLogger(logger)
+    val iterator = new SetMateInfoIterator(in.iterator(), true, allowMissingMates).toIterator
+
+    for (rec <- iterator) {
+      out.addAlignment(rec)
+      progress.record(rec)
+    }
+
+    out.close()
```

Comment:
Perhaps an implicit that adds a `safelyClose()` method to `Any` that we stick in some common def?  Though we should only use that kind of `close()` on readers, since we want to know if a writer has a problem.